sql: fix super region placement in 3-region database with secondary region

spilchen opened the pull request:
When a super region was added to a 3-region database configured for region survivability, we applied zone constraints to ensure all 5 replicas were confined to the 3 defined regions. This was intended to prevent the super region from including any region outside the database.

However, the logic mishandled secondary regions. Both the primary and secondary regions receive 2 voter replicas. The zone config logic then incorrectly added 2 more replicas to an arbitrary non-primary region, regardless of whether it was already the secondary.

This resulted in an invalid configuration: 2 replicas in the primary, 2 in the secondary, and 2 in another region. This totals 6 replicas, which exceeds the allowed maximum of 5.

The fix ensures that the secondary region is correctly accounted for when setting constraints.

Fixes #150361
Release note (bug fix): Fixed invalid zone configurations when adding a super region to a 3-region database with a secondary region and region survivability. Previously, this could result in assigning more than the allowed number of replicas.

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150413)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/regions/region_util.go` line 343 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150413#-OVNwLgt3A1vQWdqEaNk:-OVNwLgt3A1vQWdqEaNl:bt90dmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9817f24b2dee9ca0b68f4acb19d73f45334841/pkg/sql/regions/region_util.go#L343)):*
> ```Go
> 		// - There are exactly 3 regions, and
> 		// - Either there is no secondary region,
> 		//   OR the affinity region is the secondary region.
> ```

i may have had the wrong understanding, but i thought the "affinity region" is the same thing as the primary region. do you know what the difference is?
___
*[`pkg/sql/regions/region_util.go` line 357 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150413#-OVO-4Ic746Cwd0u5tpk:-OVO-4Id2fWubbUCe8Lm:b-28kivw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9817f24b2dee9ca0b68f4acb19d73f45334841/pkg/sql/regions/region_util.go#L357)):*
> ```Go
> 		//
> 		// See: https://github.com/cockroachdb/cockroach/issues/63617 for more.
> 		hasSecondary := regionConfig.HasSecondaryRegion()
> ```

do we need the `hasSecondary` variable? it seems like if that's false, then `SecondaryRegion()` will return an empty string, so all the `region == secondaryRegion` will not be true anyway.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: fix super region placement in 3-region database with secondary region

When a super region was added to a 3-region database configured for
region survivability, we applied zone constraints to ensure all 5
replicas were confined to the 3 defined regions. This was intended to
prevent the super region from including any region outside the database.

However, the logic mishandled secondary regions. Both the primary and
secondary regions receive 2 voter replicas. The zone config logic then
incorrectly added 2 more replicas to an arbitrary non-primary region,
regardless of whether it was already the secondary.

This resulted in an invalid configuration: 2 replicas in the primary, 2
in the secondary, and 2 in another region. This totals 6 replicas, which
exceeds the allowed maximum of 5.

The fix ensures that the secondary region is correctly accounted for
when setting constraints.

Fixes #150361
Release note (bug fix): Fixed invalid zone configurations when adding a super
region to a 3-region database with a secondary region and region survivability.
Previously, this could result in assigning more than the allowed number of
replicas.

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150413)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/regions/region_util.go` line 343 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150413#-OVNwLgt3A1vQWdqEaNk:-OVOINzA4zrP9MEbB85x:bctc6kz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9817f24b2dee9ca0b68f4acb19d73f45334841/pkg/sql/regions/region_util.go#L343)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i may have had the wrong understanding, but i thought the "affinity region" is the same thing as the primary region. do you know what the difference is?
</blockquote></details>

Yeah, this threw me for a loop too. The affinity region can be the secondary region for RBT tables defined in the secondary. There is an example in the secondary_region logic test:

1. We have this table:
```
statement ok
CREATE TABLE db.rbt_in_ca_central() LOCALITY REGIONAL BY TABLE IN "ca-central-1";
```

2. The primary/secondary regions are defined as follows:
```
query TT
SELECT
  region,
  CASE
    WHEN "primary" THEN 'primary'
    WHEN "secondary" THEN 'secondary'
    ELSE 'other'
  END AS role
FROM [SHOW REGIONS FROM DATABASE db]
ORDER BY region;
----
ap-southeast-2  primary
ca-central-1    secondary
us-east-1       other
```

3. The table zone config for this table is defined like this:
```
query TT
SHOW ZONE CONFIGURATION FOR TABLE db.rbt_in_ca_central
----
TABLE db.public.rbt_in_ca_central  ALTER TABLE db.public.rbt_in_ca_central CONFIGURE ZONE USING
                                     range_min_bytes = 134217728,
                                     range_max_bytes = 536870912,
                                     gc.ttlseconds = 14400,
                                     num_replicas = 5,
                                     num_voters = 5,
                                     constraints = '{+region=ap-southeast-2: 2, +region=ca-central-1: 1, +region=us-east-1: 1}',
                                     voter_constraints = '{+region=ca-central-1: 2}',
                                     lease_preferences = '[[+region=ca-central-1]]'

```

4. To ensure the table cannot have regions outside of the super region, we had to constrain `ap-southeast-2` to 2 replicas in this loop.
___
*[`pkg/sql/regions/region_util.go` line 357 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150413#-OVO-4Ic746Cwd0u5tpk:-OVOKQWqD4Lor7BmW1tM:b-hs8bty) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9817f24b2dee9ca0b68f4acb19d73f45334841/pkg/sql/regions/region_util.go#L357)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do we need the `hasSecondary` variable? it seems like if that's false, then `SecondaryRegion()` will return an empty string, so all the `region == secondaryRegion` will not be true anyway.
</blockquote></details>

I still need to check `hasSecondary` for `shouldDoubleUp` since we need to add the extra replica if the voter constraints only contain a single region (.i.e. no secondary or affinity is the secondary region). But you are correct to point out that it isn't needed inside the loop. The region name compare is sufficient. I cleaned it up a bit based on this comment.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150413#-:-OVOSkDP8c2rSXmSDOER:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/regions/region_util.go` line 343 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150413#-OVNwLgt3A1vQWdqEaNk:-OVOSiH94rpkLXFnIUD7:bvzmq38) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9817f24b2dee9ca0b68f4acb19d73f45334841/pkg/sql/regions/region_util.go#L343)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Yeah, this threw me for a loop too. The affinity region can be the secondary region for RBT tables defined in the secondary. There is an example in the secondary_region logic test:

1. We have this table:
```
statement ok
CREATE TABLE db.rbt_in_ca_central() LOCALITY REGIONAL BY TABLE IN "ca-central-1";
```

2. The primary/secondary regions are defined as follows:
```
query TT
SELECT
  region,
  CASE
    WHEN "primary" THEN 'primary'
    WHEN "secondary" THEN 'secondary'
    ELSE 'other'
  END AS role
FROM [SHOW REGIONS FROM DATABASE db]
ORDER BY region;
----
ap-southeast-2  primary
ca-central-1    secondary
us-east-1       other
```

3. The table zone config for this table is defined like this:
```
query TT
SHOW ZONE CONFIGURATION FOR TABLE db.rbt_in_ca_central
----
TABLE db.public.rbt_in_ca_central  ALTER TABLE db.public.rbt_in_ca_central CONFIGURE ZONE USING
                                     range_min_bytes = 134217728,
                                     range_max_bytes = 536870912,
                                     gc.ttlseconds = 14400,
                                     num_replicas = 5,
                                     num_voters = 5,
                                     constraints = '{+region=ap-southeast-2: 2, +region=ca-central-1: 1, +region=us-east-1: 1}',
                                     voter_constraints = '{+region=ca-central-1: 2}',
                                     lease_preferences = '[[+region=ca-central-1]]'

```

4. To ensure the table cannot have regions outside of the super region, we had to constrain `ap-southeast-2` to 2 replicas in this loop.
</blockquote></details>

this example was helpful. thanks!


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+

spilchen commented:
This bug has existed for awhile, but I'm only going to backport to 25.1 as that is when we started to implicitly create super regions for a 3-region database (see #142838), and thus more likely to see it.

blathers backport 25.1 25.2 25.3
